File: python/67_maximum_path_sum_II.py
def maximum_path_sum(triangle: list) -> int:
    
    # build the maximum path sum from the bottom of the triangle
    # maximum path sum starting at a node is equal to the value at that node
    # plus the larger of the maximum path sums that start at its left and right children
    
    for n in range(len(triangle) - 2, -1, -1):
        for i in range(0, len(triangle[n])):
            triangle[n][i] += max(triangle[n + 1][i], triangle[n + 1][i + 1])
        
    return triangle[0][0]
    
if __name__ == "__main__":
    f = open("67_triangle.txt", 'r')
    triangle = []
    for line in f.readlines():
        triangle.append([int(num) for num in line.strip().split(" ")])
    
    print(maximum_path_sum(triangle))

File: cpp/67_maximum_path_sum_II.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cstring>

int maximum_path_sum_II(std::vector<std::vector<int>> triangle){
    for (int r = triangle.size() - 2; r >= 0; r--){
        for (int c = 0; c < triangle[r].size(); c++){
            // maximum path sum from given number is sum of number and
            // maximum of path sums through adjacent numbers on next row
            triangle[r][c] += std::max(triangle[r + 1][c], triangle[r + 1][c + 1]);
        }
    }
    // top of triangle contains maximum path sum
    return triangle[0][0];
}

int main(){
    std::ifstream file("/uploads/67_triangle.txt");
    std::string s;
    std::vector<std::vector<int>> triangle;
    
    // each row is a separate line in the file
    // tokenize each line into vector of ints
    while (std::getline(file, s)){
        std::vector<int> row;
        char *pch = std::strtok(&s[0], " ");
        while (pch != NULL){
            std::string num(pch);
            row.push_back(stoi(num));
            pch = std::strtok(NULL, " ");
        }
        triangle.push_back(row);
    }
    file.close();
    std::cout << maximum_path_sum_II(triangle) << "\n";
}
